enable application to run

Rex2002 opened the pull request:
this version of the program now runs, when not enough shapes are present. 
Still needs to be tested with shapes actually present that could be recognized

Rex2002 committed:
fix: various null pointers, array out of bounds

Rex2002 committed:
fix: potential problems with re-setting midiDevice

Rex2002 committed:
Merge remote-tracking branch 'origin/main' into fix-running

# Conflicts:
#	src/main/java/de/dhbw/Main.java
#	src/main/java/de/dhbw/ui/VideoScene.java

Rex2002 committed:
feat: connect ui and shape display

Rex2002 commented:
1. color of shape drawings
2. somehow manage to place the drawn shapes in relation to the image-view, not the whole frame
currently the shapes x and y position is considered to be relative to the whole frames x and y which leads to differences between the drawn shapes and the shapes on the frame, when the frame has not the same size as the image
3. definition of playfield
for the marker classification the playfield is averaged to be a rectangle based on the recognized field_markers, since this avoids calculating the play-times of the shapes in a non-uniform way between the "upper" bars and the "lower" bars. 
Since the playfield is drawn by connecting the playfield markers, this may lead to unclarity, especially when the playmarker (indicating the position in the bar) is also drawn
4. how should the playmarker be passed?
@MalteRichert your opinion?

MalteRichert commented:
@Rex2002 addressing your comments:
1. Color: Should be determined by Shape Type, right? So that e.g. FieldMarkers are black, SoundMarkers are blue etc. Will prioritize that in styling
2. will try to set shape position in relation to ShapePane and scale ShapePane with Image. That should do the trick
3. Imo that leaves two options: either the UI recalculates the exact PlayField (risky and inefficient) or the playField is somehow passed from ShapeProcessing. That could just be an array with four tuples of x,y-coordinates. Can we do that?
4. PlayMarker could be just an ordinary Setting with the relative position within the loop. It could be its own thing within UIMessage too, but I don't see any reason for that at this time.

Rex2002 commented:
1. noice
2. very noice
3. yes, thats possible, currently the playfield is stored in an array containing the x and y position of the upper left corner, the width, height and whether it is somehow similar to a rectangle (stored as 0/1), but the x and y's of the corners can be calculated from that
4. indeed :ok_hand:

MalteRichert commented on src/main/java/de/dhbw/Main.java:
```diff
@@ -60,7 +65,20 @@ public static void runController(){
             shapeProcessor.processShapes(markerRecognizer.getShapes(), frame.width(), frame.height());
             midiAdapter.tickMidi(clock.currentBeat, shapeProcessor.getSoundMatrix(), settings);
+            EventQueues.toUI.offer(new UIMessage(markerRecognizer.getShapes()));
             // TODO add sending shapes to UI for display
```
This `TODO` is done, right?

Rex2002 replied:
yes :+1:

Rex2002 committed:
feat: add dynamic playfield drawing

Rex2002 committed:
fix: forgot something...

Rex2002 committed:
fix: bug with playfield calculation

Rex2002 committed:
feat: add sample video mp4

Rex2002 committed:
feat: change image manipulation to openCV (but JavaFX still breaks stuff...)

Commit:
chore: clean up main

Commit:
feat(ui): add button functionality

- buttons now dispatch events upon press
- events are not handled by controller yet
- buttons are not responsive to receiving events from controller yet
- also some general cleanup in UI

MalteRichert commented on src/main/java/de/dhbw/ui/VideoScene.java:
```diff
@@ -58,68 +67,67 @@ private void handleQueue() {
-        Image image = new Image( new ByteArrayInputStream(buffer.toArray() ) );
+        Image image = new Image( new ByteArrayInputStream( buffer.toArray() ) );
 
+        currentFrame.setImage(null);
```
@Rex2002 what does this do? Isn't it redundant given the next line?

Rex2002 committed:
feat: implement setting handling on midiAdapter-side

Rex2002 committed:
feat: fix flowpane-size when text changes

Rex2002 committed:
feat: move playfield drawing to shapeProcessor

Rex2002 committed:
feat: add  metronome feature, reduce midiMessage overhead

MalteRichert committed:
feat(ui): scale coordinates & drawable panes (#6)

* feat(ui): scale coordinates to draw shapes correctly in JFX

- this only works when aspect ratio of
stage equals ratio of video input
- making this robust to stage resizing
might require wrapping some UI components
to disable resizing that doesn't comply
with aspect ratio

* feat(ui): scale shapePane, fieldPane with imgView

* fix(ui): simplify calculation of img dimensions

- previous implementation was unnecessarily complex
because I forgot that I already had all the
required data

MalteRichert commented:
@Rex2002 could this be ready for a merge? I still need to do some stuff before the initial UI skeleton is done but I think that can be a new branch.
